### PR TITLE
feat(IPINFO.IO): Create new module ipinfo.io to replace iptoasn

### PR DESCRIPTION
**Pull request by yosradirgham**
New module IPINFO.IO, where we fetch the database in json format and update observable tags in inthreatv2 accordingly

Parent task: #51938

### CODE REVIEW

File: IPInfo.IO/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Changelog
+
+All notable changes to this project will be documented in this file.
+
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
+and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
```
Comment (Darkheir) [inline IPInfo.IO/CHANGELOG.md:6]: A line should be added in the changelog

File: IPInfo.IO/README.md
Comment (Darkheir) [inline IPInfo.IO/README.md:1]: If the file is empty it can be removed

File: IPInfo.IO/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+[tool.poetry]
+name = "ipinfo-io"
+version = "0.1.0"
+description = ""
+authors = ["yosra <dirgham.yosra@gmail.com>"]
```
Comment (Darkheir) [inline IPInfo.IO/pyproject.toml:5]:
```suggestion
```

File: IPInfo.IO/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+readme = "README.md"
+packages = [{include = "ipinfo"}]
+
+[tool.poetry.dependencies]
+python = ">=3.11,<3.12"
+sekoia-automation-sdk = "^1.11.0"
+pytest = "^7.4.4"
+requests-mock = "^1.11.0"
```
Comment (Darkheir) [inline IPInfo.IO/pyproject.toml:13]: Test requirements should be moved into a `[tool.poetry.group.dev.dependencies]` section

File: IPInfo.IO/ipinfo/trigger_fetch_ipinfo_database.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+from typing import Iterator
+from iso3166 import countries
+
+import requests
+from sekoia_automation.trigger import Trigger
+
+
+class TriggerFetchIPInfoDatabase(Trigger):
+    MAX_HOUR_TAG_VALID_FOR: int = 10 * 24  # Tags are valid for 10 days
+    TOKEN = os.getenv("IPINFO_TOKEN", "42039d1a8e5800")
```
Comment (Darkheir) [inline IPInfo.IO/ipinfo/trigger_fetch_ipinfo_database.py:20]: Token should be obtained from the configuration of the module
```suggestion
```

File: IPInfo.IO/ipinfo/trigger_fetch_ipinfo_database.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+
+    database_url: str = f"https://ipinfo.io/data/free/country_asn.json.gz?token={TOKEN}"
```
Comment (Darkheir) [inline IPInfo.IO/ipinfo/trigger_fetch_ipinfo_database.py:22]: The token must be obtained from the module configuration

File: IPInfo.IO/ipinfo/trigger_fetch_ipinfo_database.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+    def datetime_to_str(date: datetime) -> str:
+        return date.strftime("%Y-%m-%dT%H:%M:%SZ")
+
+    def run(self):
+        """
+        Entrypoint of the trigger
+        """
+        self.log("Starting IPInfo.IO trigger")
+        try:
+            while True:
```
Comment (Darkheir) [inline IPInfo.IO/ipinfo/trigger_fetch_ipinfo_database.py:47]:
```suggestion
            while self.running:
```

File: IPInfo.IO/ipinfo/trigger_fetch_ipinfo_database.py [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+        # Add identity of the service to the bundle
+        location_chunk.append(self.identity)
+        work_dir = self._data_path.joinpath("ipinfo_chunks").joinpath(str(uuid.uuid4()))
+        chunk_path = work_dir.joinpath("observables.json")
+        work_dir.mkdir(parents=True, exist_ok=True)
+        with chunk_path.open("w") as fp:
+            fp.write(json.dumps(location_chunk))
+
+        directory = str(work_dir.relative_to(self._data_path))
+        file_path = str(chunk_path.relative_to(work_dir))
```
Comment (Darkheir) [inline IPInfo.IO/ipinfo/trigger_fetch_ipinfo_database.py:260]: The SDK now has helper to do this. it should be used.

See `write` method in `sekoia_automation.storage`

**Review by Darkheir — CHANGES_REQUESTED** [review]
The module manifest and trigger manifest  files are missing from the module

File: IPInfo.IO/ipinfo/trigger_fetch_ipinfo_database.py (hunk 7 above)
Comment (yosradirgham) [reply IPInfo.IO/ipinfo/trigger_fetch_ipinfo_database.py:265]: Thank you for your review, weirdly the `write` from `sekoia_automation.storage` doesn't work, it keeps running until timeout. 
This is how I use it:
```python
    write(str(chunk_path), location_chunk)
```

File: IPInfo.IO/trigger_fetch_ipinfo_database.json
Comment (Darkheir) [inline IPInfo.IO/trigger_fetch_ipinfo_database.json:1]: This file shouldn't be empty

File: IPInfo.IO/manifest.json [hunk 8]
```diff
@@ -0,0 +1,12 @@
+{
+    "configuration": {
+        "$schema": "http://json-schema.org/draft-07/schema#",
+        "title": "Arguments",
+        "type": "object"
+    },
```
Comment (Darkheir) [inline IPInfo.IO/manifest.json:6]: The token must be specified in the manifest

File: IPInfo.IO/ipinfo/trigger_fetch_ipinfo_database.py [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+import requests
+from sekoia_automation.trigger import Trigger
+
+
+class TriggerFetchIPInfoDatabase(Trigger):
+    MAX_HOUR_TAG_VALID_FOR: int = 10 * 24  # Tags are valid for 10 days
+
+    @cached_property
+    def api_token(self):
+        return self.configuration.get("IPINFO_TOKEN", "42039d1a8e5800")
```
Comment (Darkheir) [inline IPInfo.IO/ipinfo/trigger_fetch_ipinfo_database.py:23]: Should be from the module configuration also this is required, we should not specify a fallback value.

**Also you just gave you API token in a public repo. You should revoke it as soon as possible.**

File: IPInfo.IO/ipinfo/trigger_fetch_ipinfo_database.py [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+
+    database_url: str = (
+        f"https://ipinfo.io/data/free/country_asn.json.gz?token={api_token}"
+    )
```
Comment (Darkheir) [inline IPInfo.IO/ipinfo/trigger_fetch_ipinfo_database.py:27]: Should be moved inside a property and rely on `self.api_token`

**Review by Darkheir — CHANGES_REQUESTED** [review]

File: IPInfo.IO/trigger_fetch_ipinfo_database.json
Comment (yosradirgham) [reply IPInfo.IO/trigger_fetch_ipinfo_database.json:1]: Yes I know I'm working on it ^^ @Darkheir

File: IPInfo.IO/manifest.json [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+            "api_token": {
+                "description": "ipinfo.io token",
+                "type": "string"
+            }
+        },
+        "required": [
+            "api_token"
+        ],
+        "secrets": [
+            "api_key"
```
Comment (Darkheir) [inline IPInfo.IO/manifest.json:21]:
```suggestion
            "api_token"
```

File: IPInfo.IO/manifest.json [hunk 12]
```diff
@@ -0,0 +1,24 @@
+{
+    "configuration": {
+        "description": "[IPINFO.IO](https://ipinfo.io//) is a database that gives access to daily updates for IP to country and IP to ASN",
+        "name": "IPInfo.IO",
+        "uuid": "92d8bb47-7c51-445d-81de-ae04edbb6f0a",
+        "slug": "ipinfo.io",
+        "version": "0.1",
```
Comment (Darkheir) [inline IPInfo.IO/manifest.json:7]: Should not be under the `configuration` attribute

File: IPInfo.IO/pyproject.toml [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+authors = []
+description = ""
+readme = "README.md"
+packages = [{include = "ipinfo"}]
+
+[tool.poetry.dependencies]
+python = ">=3.11,<3.12"
+sekoia-automation-sdk = "^1.11.0"
+pytest = "^7.4.4"
+requests-mock = "^1.11.0"
```
Comment (Darkheir) [inline IPInfo.IO/pyproject.toml:13]: Should be moved as well

File: IPInfo.IO/trigger_fetch_ipinfo_database.json [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+                "type": "integer"
+            },
+            "chunk_size": {
+                "description": "Number of items to include in each chunk. Defaults to 10 000.",
+                "type": "integer"
+            },
+            "api_token": {
+                "description": "ipinfo.io token",
+                "type": "string"
+            }
```
Comment (Darkheir) [inline IPInfo.IO/trigger_fetch_ipinfo_database.json:22]: Already defined in the module manifest

**Review by Darkheir — CHANGES_REQUESTED** [review]

File: IPInfo.IO/trigger_fetch_ipinfo_database.json [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+    "docker_parameters": "fetch_ipinfo_database",
+    "description": "Upload the ipinfo.io database regularly in json format",
+    "arguments": {
+        "title": "Arguments",
+        "type": "object",
+        "$schema": "http://json-schema.org/draft-07/schema#",
+        "properties": {
+            "interval": {
+                "description": "Interval in hours to wait between each trigger call. Defaults to 24.",
+                "type": "integer"
```
Comment (Darkheir) [inline IPInfo.IO/trigger_fetch_ipinfo_database.json:13]:
```suggestion
                "type": "integer",
                "default": 24
```

File: IPInfo.IO/trigger_fetch_ipinfo_database.json [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+            },
+            "chunk_size": {
+                "description": "Number of items to include in each chunk. Defaults to 10 000.",
+                "type": "integer"
```
Comment (Darkheir) [inline IPInfo.IO/trigger_fetch_ipinfo_database.json:17]:
```suggestion
                "type": "integer",
                "default": 10000
```

File: IPInfo.IO/trigger_fetch_ipinfo_database.json [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+            }
+        }
+    },
+    "results": {
+        "$schema": "http://json-schema.org/draft-07/schema#",
+        "title": "Results",
+        "type": "object",
+        "properties": {
+            "file_path": {
+                "description": "Database file path",
```
Comment (Darkheir) [inline IPInfo.IO/trigger_fetch_ipinfo_database.json:27]:
```suggestion
                "description": "File path of the generated observables",
```

File: IPInfo.IO/ipinfo/trigger_fetch_ipinfo_database.py (hunk 7 above)
Comment (Darkheir) [reply IPInfo.IO/ipinfo/trigger_fetch_ipinfo_database.py:265]:
```suggestion
    file_path = write("observables.json", location_chunk)
    directory = file_path.parent.as_posix()
```

And in the `send_event`, replace `file_path=file_path` by `file_path="observables.json"`.

**Review by Darkheir — CHANGES_REQUESTED** [review]
Test coverage should be improved a bit

File: IPInfo.IO/tests/test_trigger_fetch_ipinfo_database.py [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+    with open("tests/data/country_asn.json.gz", "rb") as mock_fp:
+        request_mock.get(
+            trigger.database_url,
+            content=mock_fp.read(),
+        )
+        trigger.stop()
+        try:
+            trigger.run()
+        except Exception:
+            pytest.fail("Exception occured")
```
Comment (Darkheir) [inline IPInfo.IO/tests/test_trigger_fetch_ipinfo_database.py:199]: Not needed

```suggestion
        trigger.run()
```

**Review by Darkheir — APPROVED** [review]

**Pull request merged**